### PR TITLE
Initial FM-S and FM-H support

### PR DESCRIPTION
**Pull request by Gondolindrim**
## Description

Adds support for GeonWorks' Frog Mini PCBs: FM-S and FM-H, respectively the solderable and the hotswap PCB.

## Types of Changes

- [ ] Core
- [ ] Bugfix
- [ ] New feature
- [ ] Enhancement/optimization
- [x] Keyboard (addition or update)
- [ ] Keymap/layout/userspace (addition or update)
- [ ] Documentation

## Issues Fixed or Closed by This PR

* 

## Checklist

- [x] My code follows the code style of this project: [**C**](https://docs.qmk.fm/#/coding_conventions_c), [**Python**](https://docs.qmk.fm/#/coding_conventions_python)
- [x] I have read the [**PR Checklist** document](https://docs.qmk.fm/#/pr_checklist) and have made the appropriate changes.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] I have read the [**CONTRIBUTING** document](https://docs.qmk.fm/#/contributing).
- [ ] I have added tests to cover my changes.
- [x] I have tested the changes and verified that they work and don't break anything (as well as I can manage).

### CODE REVIEW

File: keyboards/geonworks/frogmini/fmh/config.h [hunk 1]
```diff
@@ -0,0 +1,53 @@
...
+You should have received a copy of the GNU General Public License
+along with this program.  If not, see <http://www.gnu.org/licenses/>.
+*/
+
+#pragma once
+
+#include "config_common.h"
+
+/* USB Device descriptor parameter */
+#define VENDOR_ID 0x2737 // GW for GeonWorks
```
Comment (drashna) [inline keyboards/geonworks/frogmini/fmh/config.h:23]: This vendor ID appears to be in use.
https://www.the-sz.com/products/usbid/index.php?v=0x2737&p=&n=

File: keyboards/geonworks/frogmini/fms/config.h [hunk 2]
```diff
@@ -0,0 +1,58 @@
...
+ 
+You should have received a copy of the GNU General Public License 
+along with this program.  If not, see <http://www.gnu.org/licenses/>. 
+*/
+
+#pragma once
+
+#include "config_common.h"
+
+#define VENDOR_ID 0x2737 // GW for GeonWorks
```
Comment (drashna) [inline keyboards/geonworks/frogmini/fms/config.h:22]: Same.

File: keyboards/geonworks/frogmini/readme.md [hunk 3]
```diff
@@ -0,0 +1,49 @@
+# GeonWorks Frog Mini PCB firmware
+
+![FM_image](https://cdn.shopify.com/s/files/1/0477/8210/1155/files/1_600x600.jpg)
```
Comment (drashna) [inline keyboards/geonworks/frogmini/readme.md:3]: Please rehost on Imgur.

**Review by drashna — CHANGES_REQUESTED** [review]

File: keyboards/geonworks/frogmini/fmh/config.h [hunk 4]
```diff
@@ -0,0 +1,53 @@
...
+#define I2C_SCL_PAL_MODE 4
+#define I2C_SDA_PAL_MODE 4
+#define EEPROM_I2C_24LC256
+#define I2C1_CLOCK_SPEED 400000
+#define I2C1_DUTY_CYCLE FAST_DUTY_CYCLE_2
+
+#define LOCKING_SUPPORT_ENABLE
+#define LOCKING_RESYNC_ENABLE
+
+#define STM32_HSECLK 8000000
```
Comment (zvecr) [inline keyboards/geonworks/frogmini/fmh/config.h:53]:
```suggestion
```

File: keyboards/geonworks/frogmini/fmh/fmh.c [hunk 5]
```diff
@@ -0,0 +1,26 @@
...
+#include "fmh.h"
+
+void i2c_init(void) {
+    setPinInput(B6); // Try releasing special pins for a short time
+    setPinInput(B7);
+    wait_ms(10); // Wait for the release to happen
+
+    palSetPadMode(GPIOB, 6, PAL_MODE_ALTERNATE(4) | PAL_STM32_OTYPE_OPENDRAIN | PAL_STM32_PUPDR_FLOATING); // Set B6 to I2C function
+    palSetPadMode(GPIOB, 7, PAL_MODE_ALTERNATE(4) | PAL_STM32_OTYPE_OPENDRAIN | PAL_STM32_PUPDR_FLOATING); // Set B7 to I2C function
+}
```
Comment (zvecr) [inline keyboards/geonworks/frogmini/fmh/fmh.c:26]: Might not be required any more with updated board files

File: keyboards/geonworks/frogmini/fmh/mcuconf.h [hunk 6]
```diff
@@ -0,0 +1,35 @@
...
+#undef STM32_PLLQ_VALUE
+#undef STM32_PPRE1
+#undef STM32_PPRE2
+
+#define STM32_PLLM_VALUE                    4
+#define STM32_PLLN_VALUE                    168
+#define STM32_PLLP_VALUE                    4
+#define STM32_PLLQ_VALUE                    7
+#define STM32_PPRE1                         STM32_PPRE1_DIV2
+#define STM32_PPRE2                         STM32_PPRE2_DIV1
```
Comment (zvecr) [inline keyboards/geonworks/frogmini/fmh/mcuconf.h:32]:
```suggestion
```

File: keyboards/geonworks/frogmini/fmh/rules.mk [hunk 7]
```diff
@@ -0,0 +1,25 @@
+MCU = STM32F401
+
```
Comment (zvecr) [inline keyboards/geonworks/frogmini/fmh/rules.mk:2]:
```suggestion
BOARD = GENERIC_STM32_F401XC

```

File: keyboards/geonworks/frogmini/fms/config.h [hunk 8]
```diff
@@ -0,0 +1,58 @@
...
+#define I2C_SCL_PAL_MODE 4
+#define I2C_SDA_PAL_MODE 4
+#define EEPROM_I2C_24LC256
+#define I2C1_CLOCK_SPEED 400000
+#define I2C1_DUTY_CYCLE FAST_DUTY_CYCLE_2
+
+#define LOCKING_SUPPORT_ENABLE
+#define LOCKING_RESYNC_ENABLE
+
+#define STM32_HSECLK 8000000
```
Comment (zvecr) [inline keyboards/geonworks/frogmini/fms/config.h:58]:
```suggestion
```

File: keyboards/geonworks/frogmini/fms/mcuconf.h [hunk 9]
```diff
@@ -0,0 +1,38 @@
...
+#undef STM32_PLLQ_VALUE
+#undef STM32_PPRE1
+#undef STM32_PPRE2
+
+#define STM32_PLLM_VALUE                    4
+#define STM32_PLLN_VALUE                    168
+#define STM32_PLLP_VALUE                    4
+#define STM32_PLLQ_VALUE                    7
+#define STM32_PPRE1                         STM32_PPRE1_DIV2
+#define STM32_PPRE2                         STM32_PPRE2_DIV1
```
Comment (zvecr) [inline keyboards/geonworks/frogmini/fms/mcuconf.h:32]:
```suggestion
```

File: keyboards/geonworks/frogmini/fms/rules.mk [hunk 10]
```diff
@@ -0,0 +1,26 @@
+MCU = STM32F401
+
```
Comment (zvecr) [inline keyboards/geonworks/frogmini/fms/rules.mk:2]:
```suggestion
BOARD = GENERIC_STM32_F401XC

```

File: keyboards/geonworks/frogmini/fms/fms.c [hunk 11]
```diff
@@ -0,0 +1,32 @@
...
+
+void i2c_init(void) {
+    setPinInput(B6); // Try releasing special pins for a short time
+    setPinInput(B7);
+    wait_ms(10); // Wait for the release to happen
+
+    palSetPadMode(GPIOB, 6, PAL_MODE_ALTERNATE(4) | PAL_STM32_OTYPE_OPENDRAIN | PAL_STM32_PUPDR_FLOATING); // Set B6 to I2C function
+    palSetPadMode(GPIOB, 7, PAL_MODE_ALTERNATE(4) | PAL_STM32_OTYPE_OPENDRAIN | PAL_STM32_PUPDR_FLOATING); // Set B7 to I2C function
+}
+
```
Comment (zvecr) [inline keyboards/geonworks/frogmini/fms/fms.c:27]: Potentially same as `keyboards/geonworks/frogmini/fmh/fmh.c` and can be removed

**Review by zvecr — CHANGES_REQUESTED** [review]

File: keyboards/geonworks/frogmini/fmh/config.h (hunk 1 above)
Comment (Gondolindrim) [reply keyboards/geonworks/frogmini/fmh/config.h]: Updated for the lower case `0x67777` which seems not to be in use in commit [e2c6eaf](https://github.com/qmk/qmk_firmware/pull/17765/commits/e2c6eafde2028d4d32f7fa1fbe45eb1aabece8f3).

File: keyboards/geonworks/frogmini/fms/config.h (hunk 2 above)
Comment (Gondolindrim) [reply keyboards/geonworks/frogmini/fms/config.h]: Updated for the lower case `0x67777` which seems not to be in use in commit [e2c6eaf](https://github.com/qmk/qmk_firmware/pull/17765/commits/e2c6eafde2028d4d32f7fa1fbe45eb1aabece8f3).

File: keyboards/geonworks/frogmini/fms/fms.c (hunk 11 above)
Comment (Gondolindrim) [reply keyboards/geonworks/frogmini/fms/fms.c]: Committed in [467b990](https://github.com/qmk/qmk_firmware/pull/17765/commits/467b990b90bb8d4de64eebe48136352a12dfaf20)

File: keyboards/geonworks/frogmini/fms/rules.mk (hunk 10 above)
Comment (Gondolindrim) [reply keyboards/geonworks/frogmini/fms/rules.mk:2]: Committed in [467b990](https://github.com/qmk/qmk_firmware/pull/17765/commits/467b990b90bb8d4de64eebe48136352a12dfaf20)

File: keyboards/geonworks/frogmini/fms/mcuconf.h (hunk 9 above)
Comment (Gondolindrim) [reply keyboards/geonworks/frogmini/fms/mcuconf.h]: Committed in [467b990](https://github.com/qmk/qmk_firmware/pull/17765/commits/467b990b90bb8d4de64eebe48136352a12dfaf20)

File: keyboards/geonworks/frogmini/fms/config.h (hunk 8 above)
Comment (Gondolindrim) [reply keyboards/geonworks/frogmini/fms/config.h]: Committed in [467b990](https://github.com/qmk/qmk_firmware/pull/17765/commits/467b990b90bb8d4de64eebe48136352a12dfaf20)

File: keyboards/geonworks/frogmini/fmh/rules.mk (hunk 7 above)
Comment (Gondolindrim) [reply keyboards/geonworks/frogmini/fmh/rules.mk:3]: Committed in [467b990](https://github.com/qmk/qmk_firmware/pull/17765/commits/467b990b90bb8d4de64eebe48136352a12dfaf20)

File: keyboards/geonworks/frogmini/fmh/mcuconf.h (hunk 6 above)
Comment (Gondolindrim) [reply keyboards/geonworks/frogmini/fmh/mcuconf.h]: Committed in [467b990](https://github.com/qmk/qmk_firmware/pull/17765/commits/467b990b90bb8d4de64eebe48136352a12dfaf20)

File: keyboards/geonworks/frogmini/fmh/fmh.c (hunk 5 above)
Comment (Gondolindrim) [reply keyboards/geonworks/frogmini/fmh/fmh.c]: Committed in [467b990](https://github.com/qmk/qmk_firmware/pull/17765/commits/467b990b90bb8d4de64eebe48136352a12dfaf20)

File: keyboards/geonworks/frogmini/fmh/config.h (hunk 4 above)
Comment (Gondolindrim) [reply keyboards/geonworks/frogmini/fmh/config.h]: Committed in [467b990](https://github.com/qmk/qmk_firmware/pull/17765/commits/467b990b90bb8d4de64eebe48136352a12dfaf20)

File: keyboards/geonworks/frogmini/readme.md (hunk 3 above)
Comment (Gondolindrim) [reply keyboards/geonworks/frogmini/readme.md:3]: Comitted in b2169c2b8d

File: keyboards/geonworks/frogmini/readme.md [hunk 12]
```diff
@@ -0,0 +1,49 @@
+# GeonWorks Frog Mini PCB firmware
+
+![FM_image](https://i.imgur.com/mhCsc6M.jpg)
```
Comment (drashna) [inline keyboards/geonworks/frogmini/readme.md:3]: This uses smaller, but still high res version of the image that is more suitable for github markdown rendering.
```suggestion
![FM_image](https://i.imgur.com/mhCsc6Mh.jpg)
```

Comment (Gondolindrim) [reply keyboards/geonworks/frogmini/readme.md]: Committed in [c8f5183](https://github.com/qmk/qmk_firmware/pull/17765/commits/c8f518338dcbf4444355bd9fa3f72710a05d20ef)

**Review by drashna — APPROVED** [review]

**Review by fauxpark — CHANGES_REQUESTED** [review]
Missing info.json.

**Comment by Gondolindrim** [comment]
> Missing info.json.

Added in d7c3529be7

File: keyboards/geonworks/frogmini/fmh/info.json [hunk 13]
```diff
@@ -0,0 +1,71 @@
+{
+    "keyboard_name": "Frog Mini FM-H", 
+    "url": "https://geon.works", 
+    "maintainer": "gondolindrim",
```
Comment (fauxpark) [inline keyboards/geonworks/frogmini/fmh/info.json:4]:
```suggestion
    "keyboard_name": "Frog Mini FM-H", 
    "manufacturer": "GeonWorks",
    "url": "https://geon.works", 
    "maintainer": "gondolindrim", 
    "usb": {
        "vid": "0x6777",
        "pid": "0x2D28",
        "device_version": "0.0.1"
    },
```

File: keyboards/geonworks/frogmini/fmh/config.h [hunk 14]
```diff
@@ -0,0 +1,51 @@
...
+
+#include "config_common.h"
+
+/* USB Device descriptor parameter */
+#define VENDOR_ID 0x6777 // gw for geonworks
+#define PRODUCT_ID 0x2D28 // MH for "mini-H"
+#define DEVICE_VER 0x0001 // Rev. Alpha
+#define MANUFACTURER GeonWorks
+#define PRODUCT FrogMini FM-H
+
```
Comment (fauxpark) [inline keyboards/geonworks/frogmini/fmh/config.h:28]:
```suggestion
```

File: keyboards/geonworks/frogmini/fms/info.json [hunk 15]
```diff
@@ -0,0 +1,75 @@
+{
+    "keyboard_name": "Frog Mini FM-S", 
+    "url": "https://geon.works/", 
+    "maintainer": "gondolindrim",
```
Comment (fauxpark) [inline keyboards/geonworks/frogmini/fms/info.json:4]:
```suggestion
    "keyboard_name": "Frog Mini FM-S", 
    "manufacturer": "GeonWorks",
    "url": "https://geon.works/", 
    "maintainer": "gondolindrim", 
    "usb": {
        "vid": "0x6777",
        "pid": "0x2D33",
        "device_version": "0.0.1",
    },
```

File: keyboards/geonworks/frogmini/fms/config.h [hunk 16]
```diff
@@ -0,0 +1,56 @@
...
+#pragma once
+
+#include "config_common.h"
+
+#define VENDOR_ID 0x6777 // gw for geonworks
+#define PRODUCT_ID 0x2D33 // MS for "mini-S"
+#define DEVICE_VER 0x0001 // Rev. Alpha
+#define MANUFACTURER GeonWorks
+#define PRODUCT FrogMini FM-S
+
```
Comment (fauxpark) [inline keyboards/geonworks/frogmini/fms/config.h:27]:
```suggestion
```

**Review by fauxpark — CHANGES_REQUESTED** [review]

**Review by fauxpark — APPROVED** [review]

**Pull request merged**